src/main/kotlin/com/mdrobnak/lalrpop/psi/ext/LpTypeRef.kt

```diff
@@ -0,0 +1,71 @@
+import com.mdrobnak.lalrpop.psi.*
+import com.mdrobnak.lalrpop.psi.util.isRefMut
+import com.mdrobnak.lalrpop.psi.util.lifetimeOrInfer
+import org.rust.lang.core.psi.ext.childrenWithLeaves
+import org.rust.lang.core.psi.ext.elementType
+
+abstract class LpTypeRefMixin(node: ASTNode) : ASTWrapperPsiElement(node), LpTypeRef {
+    override fun resolveType(arguments: List<NonterminalGenericArgument>): String {
+        return when (childrenWithLeaves.first().elementType) {
+            // tuple
+            LpElementTypes.LPAREN -> this.typeRefList.joinToString(prefix = "(", postfix = ")") {
+                it.resolveType(arguments)
+            }
+            // array
+            LpElementTypes.LBRACKET -> "[" + this.typeRefList.first().resolveType(arguments) + "]"
+            // FIXME: not sure if this is supposed to be the type of another symbol,
+            //  but would assume so given the context requires a type while a symbol is passed.
+            LpElementTypes.POUND -> this.symbol?.resolveType(arguments) ?: "()"
+
+            // reference
+            LpElementTypes.AND -> "&${this.lifetimeOrInfer} ${if (this.isRefMut) "mut" else ""} ${
+                this.typeRefList[0].resolveType(arguments)
+            }"
+            // type
+            LpElementTypes.PATH -> {
+                val path = this.path!!
+                (path.pathId?.resolveType(arguments) ?: path.text) +
+                        when (val genericArguments = this.typeGenericArguments) {
+                            null -> ""
+                            else -> (genericArguments.lifetimeRuleList.map { it.text } +
+                                    genericArguments.typeRefList.map { it.resolveType(arguments) })
+                                .joinToString(prefix = "<", postfix = ">") { it }
+                        }
+            }
+
+            LpElementTypes.DYN -> {
+                // trait or function-like type?
+
+                when (val secondChild = this.childrenWithLeaves.elementAt(1)) {
+                    is LpPath -> {
+                        // trait
+                        val params = this.childrenWithLeaves.elementAtOrNull(2)
+                        if (params == null)
+                            "dyn " + secondChild.text
+                        else {
+                            val typeGenericArguments = params as LpTypeGenericArguments
+                            "dyn " + secondChild.text + (
+                                    typeGenericArguments.lifetimeRuleList.map { it.text } +
+                                            typeGenericArguments.typeRefList.map { it.resolveType(arguments) }
+                                    ).joinToString(prefix = "<", postfix = ">", separator = ", ") { it }
+                        }
+                    }
+                    is LpForall -> {
+                        // function-like?
+                        // FIXME: have no idea what this means
+
+                        "()"
```

Comment:
This is a Higher-Rank Trait Bounds (HRTB) type, which is usually a function but here's some more info:
https://www.reddit.com/r/rust/comments/6uobit/fora_lifetime_syntax/

I think you should be able to pass through the names and lifetimes like in the other cases.

Reply:
So the path defined [here](https://github.com/Mcat12/intellij-lalrpop/blob/master/src/main/grammars/LalrpopParser.bnf#L129) is supposed to be `Fn`/`FnMut`/`FnOnce`?

Reply:
It looks like that is what LALRPOP expects (since this grammar is adapted from the original grammar), so we can assume that. I think generally though, the forall can be used on more types than just function traits.